tests/virtualization/universal/waitfor_guests.pm

```diff
@@ -15,118 +15,38 @@ use warnings;
-        upload_logs("$xml");
-
-        # Check if settings are applied correctly in the xml
-        assert_script_run("cat $xml | grep 'controller' | grep 'pcie-root-port'");
-        assert_script_run("cat $xml | grep 'controller' | grep 'pcie-to-pci-bridge'");
-
-        # Apply xml settings to VM. Note: They will be applied after reboot.
-        assert_script_run("virt-xml-validate $xml");
-        assert_script_run("virsh define $xml");
-
-    } elsif (is_xen_host()) {
-        # We're skipping to add an additional bridge on xen
-    } else {
-        my $msg = "Unknown machine type";
-        record_soft_failure($msg);
-    }
-}
-
-# Upload xml definitions of all guests
-sub upload_machine_definitions {
-    foreach my $guest (keys %virt_autotest::common::guests) {
-        my $xml = "$guest.xml";
-        assert_script_run("virsh dumpxml $guest | tee $xml");
-        upload_logs("$xml");
-    }
-}
-
 sub run {
     my $self = shift;
     select_console('root-console');
+    my @guests = keys %virt_autotest::common::guests;
     # Fill the current pairs of hostname & address into /etc/hosts file
     assert_script_run 'virsh list --all';
-    add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (keys %virt_autotest::common::guests);
+    add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (@guests);
     assert_script_run "cat /etc/hosts";
 
     # Wait for guests to announce that installation is complete
     script_retry("test -d /tmp/guests_ip", retry => 15, delay => 120);
-    foreach my $guest (keys %virt_autotest::common::guests) {
+    foreach my $guest (@guests) {
         script_retry("test -f /tmp/guests_ip/$guest", retry => 20, delay => 120);
         record_info("$guest installed", "Guest installation completed");
     }
     record_info("All guests installed", "Guest installation completed");
-
-    # Adding the PCI bridges requires the guests to be shutdown
-    record_info("shutdown guests", "Shutting down all guests");
-    shutdown_guests();
-
-    ## Add a PCIe root port and a PCIe to PCI bridge for hotplugging
-    add_pci_bridge("$_") foreach (keys %virt_autotest::common::guests);
-
-    ## Ensure the reboot policy is set to 'restart'. This needs to happen on shutdown guest
-    # Do a shutdown and start here because some guests might not reboot because of the on_reboot=destroy policy
-    ensure_reboot_policy("$_") foreach (keys %virt_autotest::common::guests);
-    upload_machine_definitions();
-
-    record_info("Starting guests", "Starting all guests");
-    start_guests();
-
-    # Check that guests are online so we can continue and setup them
-    ensure_online $_, skip_ssh => 1, ping_delay => 45 foreach (keys %virt_autotest::common::guests);
-
-    # All guests should be now installed and running
+    my $host_version = get_var("VERSION") =~ s/-SP/./r;
+    if ($host_version > 15 && get_var("KVM")) {
```

Comment:
is_sle(>=15) or is_sle('15+') may be better ?

Reply:
Good question:
As you know, OpenQA provides Vars and store much data in them. It's simple and clean to retrieve data from Vars compared to retrieving from SUT.  In "is_sle" case, It has to login and do some operations on SUT to get data. Those operations are actually not relevant to tests.

Reply:
Really ? Which part is "not relevant" part ? Source code copied over as below:
```
sub is_sle {
    my $query = shift;
    my $version = shift // get_var('VERSION');

    return 0 unless check_var('DISTRI', 'sle');
    return 1 unless $query;

    # Version check
    return check_version($query, $version, qr/\d{2}(?:-sp\d)?/);
}

sub check_version {
    my $query = lc shift;
    my $pv = lc shift;
    my $regex = shift // qr/[^<>=+]+/;

    # Matches operator($op), version($qv), plus($plus) - regex101.com to debug ;)
    if ($query =~ /^(?(?!.*\+$)(?<op>[<>=]|[<>]=))(?<qv>$regex)(?<plus>\+)?$/i) {
        my $qv = $+{qv} or die "Query $query not matching required format $regex";

        # Compare versions if they can be parsed
        if (is_lax($pv) && is_lax($qv)) {
            $pv = version->declare($pv);
            $qv = version->declare($qv);
        }
        return $pv ge $qv if $+{plus} || $+{op} eq '>=';
        return $pv le $qv if $+{op} eq '<=';
        return $pv gt $qv if $+{op} eq '>';
        return $pv lt $qv if $+{op} eq '<';
        return $pv eq $qv if $+{op} eq '=';
    }
    # Version should be matched and processed by now
    croak "Unsupported version parameter for check_version: '$query'";
}
```

Reply:
Ah! 
I remember that it logins SUT and read "/etc/os-release" to get info?  Was the "is_sle" code improved?  Maybe, my memory was wrong.

Reply:
It is just one of the basic APIs in lib/version_utils.pm being used widely.